boot/seal.go

```diff
@@ -93,8 +106,46 @@ func sealKeyToModeenv(key, saveKey secboot.EncryptionKey, model *asserts.Model,
 	return sealKeyToModeenvUsingSecboot(key, saveKey, model, modeenv)
 }
 
+func runKeySealRequests(key secboot.EncryptionKey) []secboot.SealKeyRequest {
+	return []secboot.SealKeyRequest{
+		{
+			Key:     key,
+			KeyFile: filepath.Join(InitramfsBootEncryptionKeyDir, "ubuntu-data.sealed-key"),
+		},
+	}
+}
+
+func fallbackKeySealRequests(key, saveKey secboot.EncryptionKey) []secboot.SealKeyRequest {
+	return []secboot.SealKeyRequest{
+		{
+			Key:     key,
+			KeyFile: filepath.Join(InitramfsSeedEncryptionKeyDir, "ubuntu-data.recovery.sealed-key"),
+		},
+		{
+			Key:     saveKey,
+			KeyFile: filepath.Join(InitramfsSeedEncryptionKeyDir, "ubuntu-save.recovery.sealed-key"),
+		},
+	}
+}
+
 func sealKeyToModeenvUsingFdeSetupHook(key, saveKey secboot.EncryptionKey, model *asserts.Model, modeenv *Modeenv) error {
-	return fmt.Errorf("cannot use fde-setup hook yet")
+	// TODO: support full boot chains
+
+	for _, skr := range append(runKeySealRequests(key), fallbackKeySealRequests(key, saveKey)...) {
+		params := &FDESetupHookParams{
+			Key:     skr.Key,
+			KeyName: filepath.Base(skr.KeyFile),
+			Model:   model,
+		}
+		sealedKey, err := RunFDESetupHook("initial-setup", params)
+		if err != nil {
+			return fmt.Errorf("cannot run fde-setup hook for %s: %v", params.KeyName, err)
```

Comment:
won't err repr have already this kind of info? is just a bout the keyName. this risks to be fairly repetitive

Reply:
Nice catch! Indeed the errors from deviceManager always include enough context that it's fine to just return "err" here. I updated the code and added a test.